hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/payload/ExpressionPayload.scala

```diff
@@ -214,11 +214,7 @@ class ExpressionPayload(record: GenericRecord,
    * Init the table schema.
    */
   private def initWriteSchemaIfNeed(properties: Properties): Unit = {
-    if (writeSchema == null) {
-      ValidationUtils.checkArgument(properties.containsKey(HoodieWriteConfig.WRITE_SCHEMA.key),
-        s"Missing ${HoodieWriteConfig.WRITE_SCHEMA.key}")
-      writeSchema = new Schema.Parser().parse(properties.getProperty(HoodieWriteConfig.WRITE_SCHEMA.key))
-    }
+    writeSchema = setWriteSchema(properties)
   }
```

Comment:
Seems you want to make the schema object as singleton for different payload instance, then the null check may still makes sense here. I mean the decision `if (writeSchema == null)` should be kept.